In [53]:
import os
from decouple import Config, RepositoryEnv
config = Config(RepositoryEnv(".env"))

In [ ]:
import pandas as pd

# Specify the path to your CSV file
csv_file_path = '../data/cases.csv'  # Replace with the actual path

# Read the CSV file into a DataFrame
df = pd.read_csv(csv_file_path)

# Display the first few rows
df.head()  # By default, this shows the first 5 rows


### Connecting to the DB

In [107]:
from sshtunnel import SSHTunnelForwarder
import requests
import json
import os
import psycopg
import json
def get_db_connection():
    # Setting up the SSH tunnel with tunnel credentials
    REMOTE_HOST = config("REMOTE_HOST")
    REMOTE_SSH_PORT = int(config("REMOTE_SSH_PORT"))
    PORT = int(config("PORT"))
    SSH_KEYFILE = config("SSH_KEYFILE")
    SSH_USERNAME =  config("SSH_USERNAME")

    server = SSHTunnelForwarder(
        ssh_address_or_host=(REMOTE_HOST, REMOTE_SSH_PORT),
        ssh_username= SSH_USERNAME,
        ssh_pkey=SSH_KEYFILE,
        remote_bind_address=('localhost', PORT)
    )
    server.start()
    print("server connected")

    conn_str = f"dbname=postgres host=localhost port={server.local_bind_port} user=postgres password={config('DB_PASSWORD')}"
    conn_str_formatted = f"postgresql://postgres:{config('DB_PASSWORD')}@localhost:{server.local_bind_port}/postgres"
    return conn_str_formatted, conn_str, psycopg.connect(conn_str)

conn_str_formatted, conn_str, conn = get_db_connection()

server connected


In [108]:
import pandas as pd
import json

def exec(query, params=()):
    with conn.cursor() as cur:
        cur.execute(query, params)
        results = cur.fetchall()
        results_df = pd.DataFrame(results, columns=[desc[0] for desc in cur.description])
        return results_df

### Experiment 1

In [40]:
df = exec("""SELECT * FROM get_vector_pagerank_rrf_cases('Water leaking into the apartment from the floor above.',
                50, 50);""")
df

score  pagerank_rank       id  vector_rank  \
0   0.03088619624613922547              9   615468            1   
1   0.02985739750445632799              6   782330            8   
2   0.02815814850530376085              1   481657           25   
3   0.02763772175536881420             25  4938756            3   
4   0.02742537313432835821              7  1086651           20   
5   0.02728937728937728938              5   552773           24   
6   0.02724014336917562724              2   561149           30   
7   0.02724014336917562724             30   674990            2   
8   0.02711925287356321839             27  5041745            4   
9   0.02705156136528685548              8  1127960           21   
10  0.02683013503909026297             24   594079            7   
11  0.02674273489035478695             11  2601920           19   
12  0.02651515151515151516             28  1346648            6   
13  0.02637768817204301075              4  1186056           33   
14  0.02564935064935064935             10   772283           28   
15  0.02564102564102564102             18   849973           18   
16  0.02548562548562548563             39  1017660            5   
17  0.02504732779962137761              3  1127907           49   
18  0.02490942028985507247             36   838633            9   
19  0.02468764112599729038             31   558730           13   
20  0.02450252450252450252             14   591482           31   
21  0.02448126586057620540             17  2525134           27   
22  0.02419816723940435281             12  4953587           37   
23  0.02388842861088097211             42   685636           11   
24  0.02380952380952380952             45   798646           10   
25  0.02373595505617977528             20  2580977           29   
26  0.02350255170561375235             13  4933418           42   
27  0.02341450361252341450             41   842483           14   
28  0.02333333333333333333             15   999494           40   
29  0.02323468328141225338             47   859403           12   
30  0.02307489451476793249             19  1279441           36   
31  0.02305888483585200625             16  1095193           41   
32  0.02298701298701298701             40   566840           17   
33  0.02283341982355993773             34  1036909           22   
34  0.02277327935222672064             44  1043159           16   
35  0.02254976064499874023             21   634444           38   
36  0.02250764525993883792             49   869848           15   
37  0.02249747219413549040             26  1034620           32   
38  0.02187425292852020081             29  2532786           34   
39  0.02148215503523528075             23  4912975           46   
40  0.02148215503523528075             46   693442           23   
41  0.02139588100686498856             32  1008408           35   
42  0.02128603104212860311             22   828223           50   
43  0.02088716623600344531             48  1005731           26   
44  0.02041028845152556493             37  4920250           39   
45  0.02027649769585253456             33   867981           45   
46  0.01991281949673073112             38   960808           43   
47  0.01978557504873294347             35  5008594           48   
48  0.01932412247946228528             43   802199           44   
49  0.01843670348343245540             50   645120           47   

                                                 abbr              pagerank  \
0                          Le Vette v. Hardman Estate    0.8718440727625433   
1                                   DeHoney v. Gjarde    0.8904329943560416   
2                    Swanson v. White & Bollard, Inc.                  None   
3                              Stevens v. King County    0.6500680155483969   
4                                      Bach v. Sarich    0.8888311635157108   
5                             Hockersmith v. Ferguson    0.8997428035379669   
6                              Wood v. City of Tac

In [33]:
data = df.head(20)['data'].to_json()
data

'{"0":{"id":615468,"name":"Mabel Le Vette, Appellant, v. Hardman Estate et al., Respondents","court":{"id":9029,"name":"Washington Supreme Court","name_abbreviation":"Wash."},"analysis":{"sha256":"4361853e35bf3c3ccc332cec2f938ceae6f3b96b8e3b83e31725116a3a7d6386","simhash":"1:ad47d891b7f6505d","pagerank":{"raw":0.00000033,"percentile":0.8718440728},"char_count":10485,"word_count":1835,"cardinality":527,"ocr_confidence":0.537},"casebody":{"judges":[],"parties":["Mabel Le Vette, Appellant, v. Hardman Estate et al., Respondents."],"opinions":[{"text":"Morris, J.\\nAppeal from an order of nonsuit and dismissal, in an action brought by a tenant to recover damages for injuries to her goods, caused by leakage of water from an upper story. The facts, so far as they are pertinent to our inquiry, are about these: The Hardman Estate is the owner of a building on Yesler Way, in Seattle, the lower portion of which is divided into storerooms, and the upper is used as a hotel. Appellant, who was engag

In [53]:
from openai import AzureOpenAI

# Initialize Azure OpenAI client with key-based authentication
client = AzureOpenAI(
    azure_endpoint = config("AZURE_OPENAI_ENDPOINT"),
    api_key = config("AZURE_OPENAI_KEY"),
    api_version = "2024-05-01-preview",
)

def generate(question, context):
    completion = client.chat.completions.create(
        model='gpt-4o',
        messages= [
            {
                "role": "system",
                "content": #"You are an AI assistant that helps people find information."
                    """You are an Leagal Research AI Assistant that helps people understand relevant legal cases. Answer the user's QUESTION below using provided cases in the CONTEXT section. Keep your answer grounded in the cases in the CONTEXT section. If the CONTEXT section doesn’t contain the facts to answer the QUESTION, return NONE. Identify top 2 to 3 legal principles used in the cases provided, explain each principle and list prominent cases which illustrate it.
                    """
            },
            {
                "role": "user",
                "content": f"""
                QUESTION: {question}
                CONTEXT: {str(data)}
                """
                #INSTRUCTIONS: Answer the user's QUESTION using the DOCUMENT text above. Keep your answer grounded in the facts of the DOCUMENT. If the DOCUMENT doesn’t contain the facts to answer the QUESTION, return NONE.
            }
        ],
        max_tokens=800,
        temperature=0.3,
        frequency_penalty=0,
        presence_penalty=0,
        stop=None,
        stream=False
    )
    return completion.to_json()



In [54]:
completion_json = generate(
    'Water leaking into the apartment from the floor above. What are the prominent legal precedents in Washington on this problem?',
    #'Water leaking into the apartment from the floor above. What is the legal liability of the landlord in this case?',
    data
)

print(json.loads(completion_json)['choices'][0]['message']['content'])

Based on the provided cases, here are the top 3 legal principles relevant to the issue of water leaking into an apartment from the floor above in Washington:

1. **Landlord's Duty to Repair and Maintain Safe Conditions**:
   - **Principle**: A landlord has a duty to maintain the premises in a condition of reasonable safety, especially when the premises remain under their control. If a landlord negligently fails to repair or maintain the premises, resulting in damage to a tenant, the landlord can be held liable.
   - **Prominent Cases**:
     - **Mabel Le Vette v. Hardman Estate**: The Washington Supreme Court held that the landlord was negligent in failing to control and keep the upper stories in a condition of reasonable safety, leading to water damage in the tenant's storeroom below.
     - **Martindale Clothing Co. v. Spokane & Eastern Trust Co.**: The court found that the landlord had a duty to inform the tenant of the means to prevent freezing and bursting of water pipes, and fail

### Experiment 2: Trying to incorporate sorting by pagerank (a little worse results)

In [57]:
df = exec("""SELECT * FROM get_vector_pagerank_rrf_cases('Water leaking into the apartment from the floor above.', 
                50, 50)
             ORDER BY pagerank_rank;""")
df

score  pagerank_rank       id  vector_rank  \
0   0.02815814850530376085              1   481657           25   
1   0.02724014336917562724              2   561149           30   
2   0.02504732779962137761              3  1127907           49   
3   0.02637768817204301075              4  1186056           33   
4   0.02728937728937728938              5   552773           24   
5   0.02985739750445632799              6   782330            8   
6   0.02742537313432835821              7  1086651           20   
7   0.02705156136528685548              8  1127960           21   
8   0.03088619624613922547              9   615468            1   
9   0.02564935064935064935             10   772283           28   
10  0.02674273489035478695             11  2601920           19   
11  0.02419816723940435281             12  4953587           37   
12  0.02350255170561375235             13  4933418           42   
13  0.02450252450252450252             14   591482           31   
14  0.02333333333333333333             15   999494           40   
15  0.02305888483585200625             16  1095193           41   
16  0.02448126586057620540             17  2525134           27   
17  0.02564102564102564102             18   849973           18   
18  0.02307489451476793249             19  1279441           36   
19  0.02373595505617977528             20  2580977           29   
20  0.02254976064499874023             21   634444           38   
21  0.02128603104212860311             22   828223           50   
22  0.02148215503523528075             23  4912975           46   
23  0.02683013503909026297             24   594079            7   
24  0.02763772175536881420             25  4938756            3   
25  0.02249747219413549040             26  1034620           32   
26  0.02711925287356321839             27  5041745            4   
27  0.02651515151515151516             28  1346648            6   
28  0.02187425292852020081             29  2532786           34   
29  0.02724014336917562724             30   674990            2   
30  0.02468764112599729038             31   558730           13   
31  0.02139588100686498856             32  1008408           35   
32  0.02027649769585253456             33   867981           45   
33  0.02283341982355993773             34  1036909           22   
34  0.01978557504873294347             35  5008594           48   
35  0.02490942028985507247             36   838633            9   
36  0.02041028845152556493             37  4920250           39   
37  0.01991281949673073112             38   960808           43   
38  0.02548562548562548563             39  1017660            5   
39  0.02298701298701298701             40   566840           17   
40  0.02341450361252341450             41   842483           14   
41  0.02388842861088097211             42   685636           11   
42  0.01932412247946228528             43   802199           44   
43  0.02277327935222672064             44  1043159           16   
44  0.02380952380952380952             45   798646           10   
45  0.02148215503523528075             46   693442           23   
46  0.02323468328141225338             47   859403           12   
47  0.02088716623600344531             48  1005731           26   
48  0.02250764525993883792             49   869848           15   
49  0.01843670348343245540             50   645120           47   

                                                 abbr              pagerank  \
0                    Swanson v. White & Bollard, Inc.                  None   
1                              Wood v. City of Tacoma    0.9670882356461709   
2                                      Foisy v. Wyman    0.9421446964668596   
3    Stuart v. Coldwell Banker Commercial Group, Inc.    0.9117369391033995   
4                             Hockersmith v. Ferguson    0.8997428035379669   
5                                   DeHoney v. Gjarde    0.8904329943560416   
6                                      Bach v. Sar

In [58]:
data = df.head(10)[['pagerank', 'data']].to_json()
data

'{"pagerank":{"0":null,"1":0.9670882356,"2":0.9421446965,"3":0.9117369391,"4":0.8997428035,"5":0.8904329944,"6":0.8888311635,"7":0.8825139999,"8":0.8718440728,"9":0.8667748583},"data":{"0":{"id":481657,"name":"A. L. E. Swanson, Appellant, v. White & Bollard, Inc., Respondent","court":{"id":9029,"name":"Washington Supreme Court","name_abbreviation":"Wash."},"analysis":{"sha256":"a448db74755e18330b1c583779707b6ad00a9547193873d3a0da892803174fb9","simhash":"1:26adf80d084c6556","char_count":5871,"word_count":1066,"cardinality":358,"ocr_confidence":0.485},"casebody":{"judges":[],"parties":["A. L. E. Swanson, Appellant, v. White & Bollard, Inc., Respondent."],"opinions":[{"text":"Geraghty, J.\\nThis is an appeal from a judgment of dismissal entered after the court had sustained a challenge to the sufficiency of the evidence and discharged the jury, at the close of appellant\\u2019s case.\\nThe appellant, with his mother, occupied an apartment in a building owned by the respondent, under a wri

In [59]:
completion_json = generate(
    'Water leaking into the apartment from the floor above. What are the prominent legal precedents in Washington on this problem?',
    data
)

print(json.loads(completion_json)['choices'][0]['message']['content'])

The prominent legal precedents in Washington regarding water leaking into an apartment from the floor above involve several key principles:

1. **Landlord's Duty to Repair and Maintain**: Generally, landlords are not obligated to repair or maintain leased premises unless explicitly stated in the lease. However, if the landlord undertakes repairs, they must do so non-negligently.

   - **Swanson v. White & Bollard, Inc.**: The landlord was not initially obligated to repair a leaking faucet, but once they undertook the repair, they were required to do so in a non-negligent manner. The court found no causal connection between the landlord's repair and the injury caused by the broken faucet handle, and the tenant had agreed to make all repairs (185 Wash. 407, 1936).

2. **Implied Warranty of Habitability**: This principle holds that landlords must ensure the premises are habitable and fit for living. This includes maintaining essential services like plumbing and preventing conditions that 

### Experiment 3: With 3 way reranker (good structure, but Faisy is missing in response)

In [19]:
df = exec("""SELECT * FROM get_vector_pagerank_rrf_rerank_cases('Water leaking into the apartment from the floor above.', 
                50, 50);""")
df

score  pagerank_rank  relevance  semantic_rank       id  \
0   0.03614247045619594639              8  -9.194025             50  1127960   
1   0.03258881553912891909             41  -9.136231             49   842483   
2   0.03416867954911433173             36  -8.636745             48   838633   
3   0.03382706025309956989             17  -8.552217             47  2525134   
4   0.03226738208771088113             34  -8.408062             46  1036909   
5   0.03372197676321387662             12  -8.134167             45  4953587   
6   0.03350381322626558749             42  -7.990444             44   685636   
7   0.03225849850907641013             21  -7.972852             43   634444   
8   0.03303860485003970436             47  -7.959902             42   859403   
9   0.03177524302753010180             29  -7.887039             41  2532786   
10  0.03711925287356321839             27  -7.814040             40  5041745   
11  0.03287428945323682165             44  -7.734824             39  1043159   
12  0.03749345892203035060              5  -7.710566             38   552773   
13  0.03338417286528339641             19  -7.682215             37  1279441   
14  0.03765681003584229391              2  -7.656595             36   561149   
15  0.03200847082470896496             46  -7.638021             35   693442   
16  0.03437425292852020081             20  -7.537093             34  2580977   
17  0.03425523987765676310             13  -7.442093             33  4933418   
18  0.03114606291324383891             33  -7.335201             32   867981   
19  0.03238489199587597757             32  -7.319906             31  1008408   
20  0.03853648424543946932              7  -7.316692             30  1086651   
21  0.03636823018845490756              3  -7.296780             29  1127907   
22  0.03700466200466200466             18  -7.295315             28   849973   
23  0.03081837535302550367             43  -7.270304             27   802199   
24  0.03486223040908742037             15  -7.261832             26   999494   
25  0.03850744077270772813             11  -7.095154             25  2601920   
26  0.03181758140149263588             38  -7.090520             24   960808   
27  0.03842588094312734810              4  -7.050404             23  1186056   
28  0.03983284370658832640             25  -6.890940             22  4938756   
29  0.03485332427228451693             49  -6.837715             21   869848   
30  0.03220062771607918880             35  -6.836582             20  5008594   
31  0.03394425889022986893             22  -6.714348             19   828223   
32  0.04006065618968844775             30  -6.713632             18   674990   
33  0.03446916802224826776             23  -6.667007             17  4912975   
34  0.03784553586283939564             31  -6.600910             16   558730   
35  0.03632034632034632034             40  -6.439123             15   566840   
36  0.03801603801603801603             14  -6.357308             14   591482   
37  0.03934798078633695072             10  -6.259530             13   772283   
38  0.04040404040404040405             28  -6.240168             12  1346648   
39  0.03957013252787900676             39  -6.099052             11  1017660   
40  0.03272241776914674111             50  -5.922514             10   645120   
41  0.03490304207471397073             37  -5.747809              9  4920250   
42  0.03851540616246498599             45  -5.656102              8   798646   
43  0.04175550817341862118             24  -5.223325              7   594079   
44  0.03830940988835725678             16  -5.097034              6  1095193   
45  0.04354276388991914547              1  -4.603706              5   481657   
46  0.03651216623600344531             48  -4.587784              4  1005731   
47  0.04573041337747220101              6  -4.413364              3   782330   
48  0.04701522850420374160              9  -3.736470              2   615468   
49  0.038890914817086

In [20]:
from openai import AzureOpenAI

# Initialize Azure OpenAI client with key-based authentication
client = AzureOpenAI(
    azure_endpoint = config("AZURE_OPENAI_ENDPOINT"),
    api_key = config("AZURE_OPENAI_KEY"),
    api_version = "2024-05-01-preview",
)

def generate2(question, context):
    completion = client.chat.completions.create(
        model='gpt-4o',
        messages= [
            {
                "role": "system",
                "content":
                    """You are an Leagal Research AI Assistant that helps people understand relevant legal cases. Answer the user's QUESTION below using provided cases in the CONTEXT section. Keep your answer grounded in the cases in the CONTEXT section. If the CONTEXT section doesn’t contain the facts to answer the QUESTION, return NONE. Identify top 2 to 3 legal principles used in the cases provided, explain each principle and list prominent cases which illustrate it. 
                    
                    """
            },
            {
                "role": "user",
                "content": f"""
                QUESTION: {question}
                CONTEXT: {str(context)}
                """
            }
        ],
        max_tokens=800,
        temperature=0.3,
        frequency_penalty=0,
        presence_penalty=0,
        stop=None,
        stream=False
    )
    return completion.choices[0].message.content

generate2(
    'Water leaking into the apartment from the floor above. What are the prominent legal precedents in Washington on this problem?',
    df.head(21)[['pagerank', 'data']].to_json()
)

"The prominent legal precedents in Washington regarding water leaking into an apartment from the floor above involve principles related to landlord liability for defects and the duty to repair. Here are the top legal principles and illustrative cases:\n\n### 1. **Landlord's Duty to Repair and Maintain Safe Conditions**\nLandlords have a duty to maintain rental properties in a safe and habitable condition. This includes addressing defects that could cause water leakage from one apartment to another.\n\n**Illustrative Cases:**\n- **Andrews v. McCutcheon (1943)**: The landlord was found liable for injuries sustained by a tenant due to a defective stairway. The court held that if the landlord retains control over common areas or facilities, they have a duty to keep them in a reasonably safe condition.\n- **Taylor v. Stimson (1958)**: The court ruled that a landlord is not responsible for defects known to the tenant at the time of leasing unless there is an agreement to repair. However, if 

### Experiment 4: Trying RRF (semantic + graph ranks) - the best so far
Reranker seems to push down irrelevant results. So this option does RRF between semantic and grapph ranks

In [79]:

df = exec("""SELECT * FROM get_vector_rerank_pagerank_rrf2_cases('Water leaking into the apartment from the floor above.', 
                50, 52);""")
df_graph = df
df

InternalError_: azure_ml.invoke: 408: upstream request timeout
Please check this guide to understand why this error code might have been returned 
https://docs.microsoft.com/en-us/azure/machine-learning/how-to-troubleshoot-online-endpoints#http-status-codes

CONTEXT:  PL/pgSQL function semantic_relevance(text,integer) line 7 at assignment
SQL statement "WITH vector AS (
		SELECT cases.id, RANK() OVER (ORDER BY description_vector <=> embedding) AS vector_rank, cases.data ->> 'name_abbreviation' AS abbr, (cases.data#>>'{analysis, pagerank, percentile}')::NUMERIC AS pagerank, cases.data
		FROM cases
		WHERE (cases.data#>>'{court, id}')::integer IN (9029) -- Washington Supreme Court (9029) or Washington Court of Appeals (8985)
		ORDER BY description_vector <=> embedding
		LIMIT consider_n
	),
	result AS (
		SELECT * 
		FROM jsonb_array_elements(
				semantic_relevance(query,
				consider_n)
			) WITH ORDINALITY AS elem(relevance)
	),
	semantic_ranked AS (
		SELECT result.relevance::DOUBLE PRECISION AS relevance, RANK() OVER (ORDER BY result.relevance::DOUBLE PRECISION DESC) AS semantic_rank, vector.*
		FROM vector
		JOIN result ON vector.vector_rank = result.ordinality
		ORDER BY relevance DESC
	),
	graph_ranked AS (
		SELECT RANK() OVER (ORDER BY semantic_ranked.pagerank DESC) AS pagerank_rank, semantic_ranked.* 
		FROM semantic_ranked ORDER BY semantic_ranked.pagerank DESC
	),
	rrf AS (
		SELECT
		    COALESCE(1.0 / (60 + graph_ranked.pagerank_rank), 0.0) +
			COALESCE(1.0 / (60 + graph_ranked.semantic_rank), 0.0) AS score,
			graph_ranked.*
		FROM graph_ranked
		ORDER BY score DESC
		LIMIT top_n
	)
	SELECT * 
	FROM rrf"
PL/pgSQL function get_vector_rerank_pagerank_rrf2_cases(text,integer,integer) line 5 at RETURN QUERY

In [36]:
from openai import AzureOpenAI

# Initialize Azure OpenAI client with key-based authentication
client = AzureOpenAI(
    azure_endpoint = config("AZURE_OPENAI_ENDPOINT"),
    api_key = config("AZURE_OPENAI_KEY"),
    api_version = "2024-05-01-preview",
)

def generate3(question, context):
    completion = client.chat.completions.create(
        model='gpt-4o',
        messages= [
            {
                "role": "system",
                "content":
                    """You are an Leagal Research AI Assistant that helps people understand relevant legal cases. Answer the user's QUESTION below using provided cases in the CONTEXT section. Keep your answer grounded in the cases in the CONTEXT section. If the CONTEXT section doesn’t contain the facts to answer the QUESTION, return NONE. Identify top 2 legal principles used in the cases provided, explain them and group prominent cases by those principles.
                    """
            },
            {
                "role": "user",
                "content": f"""
                QUESTION: {question}
                CONTEXT: {str(context)}
                """
            }
        ],
        max_tokens=800,
        temperature=0.3,
        frequency_penalty=0,
        presence_penalty=0,
        stop=None,
        stream=False
    )
    return completion.choices[0].message.content

generate3(
    'Water leaking into the apartment from the floor above. What are the prominent legal precedents in Washington on this problem?',
    df.iloc[1:21]['data'].to_json()
)

"The prominent legal precedents in Washington regarding water leaking into an apartment from the floor above are based on two main legal principles: the duty of care in maintaining premises and the implied warranty of habitability.\n\n### 1. **Duty of Care in Maintaining Premises**\nThis principle holds landlords responsible for maintaining the premises in a safe and habitable condition, especially when defects are known or should have been known to them. The landlord must ensure that the property does not pose a danger to the tenants.\n\n#### **Prominent Cases:**\n- **Martindale Clothing Co. v. Spokane & Eastern Trust Co.** (1914): The court held the landlord liable for damages caused by water leaking from a pipe in the upper story, which the landlord failed to maintain properly. The landlord's negligence in maintaining the water pipe led to the flooding of the tenant's store, causing significant damage.\n- **Thomas v. Housing Authority of the City of Bremerton** (1967): The court fou

#### Golden dataset

In [109]:
graph_exp4_dataset_ids = [782330, 615468, 1095193,1034620,772283,1186056,1127907,591482,594079,561149,1086651,2601920,552773,1346648,4912975,999494,1005731,828223,4920250,4933418,798646,]
df_graph_exp4 = exec(f"""SELECT data FROM cases WHERE id::INT IN ({",".join(map(str, graph_exp4_dataset_ids))});""")
df_graph_exp4

data
0   {'id': 772283, 'name': 'The Arnold-Evans Compa...
1   {'id': 782330, 'name': 'J. H. DeHoney et al., ...
2   {'id': 798646, 'name': 'F. W. Cordes, Responde...
3   {'id': 828223, 'name': 'Albert Boyer et al., A...
4   {'id': 999494, 'name': 'Katherine Papac, Respo...
5   {'id': 2601920, 'name': 'John Pappas, Appellan...
6   {'id': 1005731, 'name': 'John Finley et al., R...
7   {'id': 1034620, 'name': 'Hans T. Jorgensen et ...
8   {'id': 1095193, 'name': 'Carrie Thomas, Respon...
9   {'id': 1086651, 'name': 'Gerard Bach et al., R...
10  {'id': 1127907, 'name': 'Ronald D. Foisy et al...
11  {'id': 1186056, 'name': 'F. Craig Stuart, et a...
12  {'id': 4933418, 'name': 'Puget Investment Comp...
13  {'id': 4920250, 'name': 'Arthur L. Fleenor et ...
14  {'id': 4912975, 'name': 'Ida Mae Conradi, Indi...
15  {'id': 1346648, 'name': 'Vito Tombari, Respond...
16  {'id': 615468, 'name': 'Mabel Le Vette, Appell...
17  {'id': 561149, 'name': 'Victoria A. Wood, Appe...
18  {'id': 591482, 'name': 'T. Shigeta, Appellant,...
19  {'id': 552773, 'name': 'A. P. Hockersmith et a...
20  {'id': 594079, 'name': 'Martindale Clothing Co...

In [114]:
from openai import AzureOpenAI

# Initialize Azure OpenAI client with key-based authentication
client = AzureOpenAI(
    azure_endpoint = config("AZURE_OPENAI_ENDPOINT"),
    api_key = config("AZURE_OPENAI_KEY"),
    api_version = "2024-05-01-preview",
)

def generate3g(question, context):
    completion = client.chat.completions.create(
        model='gpt-4o',
        messages= [
            {
                "role": "system",
                "content":
                    """You are a Legal Research AI Assistant that helps people understand relevant legal cases. CONTEXT section below provides legal cases that are relevant to the user's QUESTION . Summarize cases in in the CONTEXT section using table format, with following columns:
                    1. ID - id of the case
                    2. Abbreviation - name_abbreviation of the case
                    3. Year - the year of the case
                    4. Summary - brief summary of the case

                    Output data in Markdown format.
                    """
            },
            {
                "role": "user",
                "content": f"""
                QUESTION: {question}
                CONTEXT: {str(context)}
                """
            }
        ],
        max_tokens=800,
        temperature=0.3,
        frequency_penalty=0,
        presence_penalty=0,
        stop=None,
        stream=False
    )
    return completion.choices[0].message.content

generate3g(
    'Water leaking into the apartment from the floor above. What are the prominent legal precedents in Washington on this problem?',
    df_graph_exp4['data'].to_json()
)

"```markdown\n| ID  | Abbreviation                       | Year | Summary                                                                                       |\n|-----|------------------------------------|------|-----------------------------------------------------------------------------------------------|\n| 0   | Arnold-Evans Co. v. Hardung        | 1925 | Tenants were held liable for the cost of a new boiler installation, as they had agreed to keep the building in good repair. |\n| 1   | DeHoney v. Gjarde                  | 1925 | Builder was found liable for damages due to defective construction of a roof, despite tenant's initial acceptance of the building. |\n| 2   | Cordes v. Guy Investment Co.       | 1927 | Landlord was held liable for damages to tenant's property caused by a leaking roof, as the lease required the landlord to make necessary repairs. |\n| 3   | Boyer v. City of Tacoma            | 1930 | City was held liable for damages caused by sewage backup into a baseme

In [115]:

f"""
                QUESTION: 'Water leaking into the apartment from the floor above. What are the prominent legal precedents in Washington on this problem?'
                CONTEXT: {str(df_graph_exp4['data'].to_json())}
                """

'\n                QUESTION: \'Water leaking into the apartment from the floor above. What are the prominent legal precedents in Washington on this problem?\'\n                CONTEXT: {"0":{"id":772283,"name":"The Arnold-Evans Company, Plaintiff, v. J. A. Hardung, Respondent, W. E. Bonza et al., Appellants","court":{"id":9029,"name":"Washington Supreme Court","name_abbreviation":"Wash."},"analysis":{"sha256":"921337bc3ab957f84289bb2d74dd6ab73c3326aa69a6e79bc0345344e641ab0b","simhash":"1:113624688c2e46da","pagerank":{"raw":0.0000003216,"percentile":0.8667748583},"char_count":10070,"word_count":1773,"cardinality":568,"ocr_confidence":0.51},"casebody":{"judges":[],"parties":["The Arnold-Evans Company, Plaintiff, v. J. A. Har-dung, Respondent, W. E. Bonza et al., Appellants."],"opinions":[{"text":"Tolman, J.\\nThe appellants, Bonza and wife, became the tenants of respondent, Hardung, April 1, 1922, occupying under an oral lease an apartment house in the city of Spokane, known as the Black

### Experiment 5: Incorporating cases from both Washington Supreme Court (9029) or Washington Court of Appeals (8985)

In [37]:
generate3(
    'Water leaking into the apartment from the floor above. What are the prominent legal precedents in Washington on this problem?',
    df.iloc[1:11]['data'].to_json()
)

"The prominent legal precedents in Washington regarding water leaking into an apartment from the floor above can be grouped under two main legal principles: **Negligence** and **Implied Warranty of Habitability**.\n\n### 1. Negligence\nNegligence involves a failure to exercise the care that a reasonably prudent person would exercise in like circumstances. In the context of landlord-tenant relationships, this principle can apply when a landlord or tenant fails to maintain the premises in a way that prevents foreseeable harm, such as water damage from leaking pipes.\n\n#### Prominent Cases:\n- **Martindale Clothing Co. v. Spokane & Eastern Trust Co.**: The court found that the landlord could be liable for damages resulting from the negligent maintenance of a water pipe that burst and caused water damage to the tenant's property. The court emphasized the landlord's duty to inform the tenant of the means to prevent such damage.\n- **Jorgensen v. Massart**: The court held that a tenant who 

In [51]:
df = exec("""SELECT * FROM get_vector_rerank_pagerank_rrf2_cases_v2('Water leaking into the apartment from the floor above.', 
                50, 50);""")
df

score  pagerank_rank  relevance  semantic_rank       id  \
0   0.03083491461100569260              8  -3.736470              2   615468   
1   0.03036576949620427882              3  -5.747809              9  1186056   
2   0.03030998851894374283              7  -4.603706              5  1127960   
3   0.03009049773755656109              5  -5.656102              8   782330   
4   0.02932363013698630137              4  -6.259530             13   552773   
5   0.02681010928961748634              1  -7.656595             36   481657   
6   0.02664576802507836991             27  -5.097034              6  2532786   
7   0.02655328011107254426             26  -5.223325              7  1346648   
8   0.02643831060857998005              2  -7.682215             37   561149   
9   0.02637768817204301075             33  -4.587784              4  1036909   
10  0.02619863013698630137             13  -6.836582             20   999494   
11  0.02599156118143459915             15  -6.714348             19  1095193   
12  0.02587301587301587302             40  -4.413364              3   591938   
13  0.02584921292460646231             25  -6.099052             11  5041745   
14  0.02582740488710176307             46  -2.403610              1  3982502   
15  0.02557875991581673952             11  -7.270304             27  4953587   
16  0.02535559678416821274              6  -7.710566             38  1086651   
17  0.02520608750792644261             16  -7.050404             23  2525134   
18  0.02514142049025769956             14  -7.261832             26  3977147   
19  0.02481203007518796992             10  -7.638021             35  2601920   
20  0.02472527472527472527             18  -7.090520             24  1279441   
21  0.02439374372219830679              9  -7.887039             41   772283   
22  0.02438672438672438672             39  -5.922514             10  5752736   
23  0.02430555555555555556             36  -6.240168             12  1017660   
24  0.02364723883711225483             19  -7.319906             31  2580977   
25  0.02358163406852545429             29  -6.837715             21   558730   
26  0.02355875831485587584             28  -6.890940             22   674990   
27  0.02351332877648667122             35  -6.667007             17  4920250   
28  0.02341269841269841270             12  -8.134167             45   591482   
29  0.02336956521739130435             20  -7.335201             32   634444   
30  0.02296181630546955624             42  -6.600910             16   685636   
31  0.02268782544011901810             49  -6.357308             14  1005731   
32  0.02242424242424242424             50  -6.439123             15   869848   
33  0.02234706616729088639             30  -7.296780             29  1855870   
34  0.02234432234432234432             45  -6.713632             18  8848167   
35  0.02207792207792207792             17  -9.194025             50   849973   
36  0.02204819277108433735             23  -7.814040             40  4938756   
37  0.02177964127649662241             21  -8.408062             46   473788   
38  0.02174169916105399976             31  -7.442093             33  1008408   
39  0.02166569598136284217             41  -7.095154             25   358512   
40  0.02161349976883957466             24  -7.972852             43  1034620   
41  0.02150786308973172988             32  -7.537093             34  1352760   
42  0.02145438121047877146             22  -8.636745             48   594079   
43  0.02142038946162657503             37  -7.316692             30   566840   
44  0.02107237422771403354             43  -7.295315             28  1043159   
45  0.02030509173366316223             38  -7.734824             39   842483   
46  0.02025368248772504091             34  -7.990444             44   838633   
47  0.01941930618401206636             44  -7.959902             42   798646   
48  0.01869158878504672898             47  -8.552217             47   693442   
49  0.018433571185864

In [43]:
generate3(
    'Water leaking into the apartment from the floor above. What are the prominent legal precedents in Washington on this problem?',
    df.iloc[0:20]['data'].to_json()
)

"The prominent legal precedents in Washington regarding water leaking into an apartment from the floor above can be grouped into two main legal principles: \n\n1. **Landlord's Duty to Repair and Maintain Safe Conditions**:\n   - **Principle**: Landlords have a duty to maintain the premises in a condition that is safe and habitable. This includes addressing issues such as water leaks that can cause damage to a tenant's property or affect the habitability of the premises.\n   - **Cases**:\n     - **Mabel Le Vette v. Hardman Estate (1914)**: The Washington Supreme Court held that a landlord is responsible for maintaining the premises in a condition of reasonable safety. If a landlord negligently fails to repair or maintain the premises, resulting in damage to a tenant's property, the landlord can be held liable for damages. In this case, the landlord was found negligent for failing to repair a broken water pipe, which caused water to leak into the tenant's storeroom, damaging her goods.\n